scripts/main.py

```diff
@@ -0,0 +1,180 @@
+            break
+        # Check convergence --> norm of diff btw x_sim and x_ref (only for first joint)
+        if convergenceCriteria(x_sim[k + 1], np.array([1, 0, 0, 1, 0, 0])):
+            convergence = 1
+            break
+    x_v = controller.getLastViableState()
+    return k, convergence, x_sim, stats, x_v
+
+
+if __name__ == '__main__':
+    args = parse_args()
+    # Define the available systems and controllers
+    available_systems = {'double_pendulum': 'DoublePendulumModel',
+                         'triple_pendulum': 'TriplePendulumModel'}
+    available_controllers = {'naive': 'NaiveController',
+                             'st': 'STController',
+                             'stwa': 'STWAController',
+                             'htwa': 'HTWAController',
+                             'receding': 'RecedingController',
+                             'abort': 'SafeBackupController'}
+
+    if args['init_conditions']:
+        args['controller'] = 'receding'
+    # Check if the system and controller selected are available
+    if args['system'] not in available_systems.keys():
+        raise ValueError('Unknown system. Available: ' + str(available_systems))
+    if args['controller'] not in available_controllers.keys():
+        raise ValueError('Unknown controller. Available: ' + str(available_controllers))
+
+    # Define the configuration object, model, simulator and controller
+    conf = Parameters(args['system'], args['controller'], args['rti'])
+    # Set the safety margin
+    conf.alpha = args['alpha']
+    model = getattr(models, available_systems[args['system']])(conf)
+    gc = GravityCompensation(conf, model)
+    simulator = SimDynamics(model)
+    controller = getattr(controllers, available_controllers[args['controller']])(simulator)
+
+    x_ref = np.array([conf.q_max - 0.05, np.pi, np.pi, 0, 0, 0])
+    controller.setReference(x_ref)
+
+    # Check if data folder exists, if not create it
+    if not os.path.exists(conf.DATA_DIR):
+        os.makedirs(conf.DATA_DIR)
+    data_name = conf.DATA_DIR + args['controller'] + '_'
+
+    # If ICs is active, compute the initial conditions for all the controller
+    if args['init_conditions']:
+        from scipy.stats import qmc
+        test_num = 500                      # Higher than the one reported in the configuration file
+        sampler = qmc.Halton(d=controller.ocp.dims.nu, scramble=False)
+        sample = sampler.random(n=test_num)
+        l_bounds = model.x_min[:model.nq] + conf.state_tol
+        u_bounds = model.x_max[:model.nq] - conf.state_tol
+        x0_vec = qmc.scale(sample, l_bounds, u_bounds)
+
+        # Soft constraints an all the trajectory
+        for i in range(controller.N):
+            controller.ocp_solver.cost_set(i, "zl", conf.ws_r * np.ones((1,)))
+        controller.ocp_solver.cost_set(controller.N, "zl", conf.ws_t * np.ones((1,)))
+
+        overall_success, x_guess_vec, u_guess_vec = init_guess(test_num)
```

Comment:
I don't get why you try to solve 500 OCP's if then you need only `conf.test_num` initial conditions. Couldn't you keep track of how many problems you have solved and stop once you've reached `conf.test_num`?